src/main/java/com/google/sps/servlets/EventFreeTimeServlet.java

```diff
@@ -0,0 +1,198 @@
+    Set<Time> busyTimesForAttendees = new TreeSet<>();
+    response.setContentType("text/html");
+
+    JsonBatchCallback<FreeBusyResponse> callback = new JsonBatchCallback<FreeBusyResponse>() {
+
+      public void onSuccess(FreeBusyResponse apiResponse, HttpHeaders responseHeaders) throws IOException {
+        FreeBusyCalendar busyCalendar = apiResponse.getCalendars().get("primary");
+        List<TimePeriod> listBusyTimes = busyCalendar.getBusy();
+        listBusyTimes.forEach(busyTime -> {
+          busyTimesForAttendees.add(new Time(busyTime.getStart().getValue(), busyTime.getEnd().getValue()));
+        });
+      }
+
+      public void onFailure(GoogleJsonError e, HttpHeaders responseHeaders) {
+        System.err.println("Error Message: " + e.getMessage());
+        response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      }
+    };
+
+    String eventId = request.getParameter("eventId");
+    if (eventId == null || eventStorageObject.getEvent(eventId) == null) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    DateTimeRange timeRange = eventStorageObject.getEvent(eventId).getDateTimeRange();
+    Long duration = eventStorageObject.getEvent(eventId).getDuration();
+    if (timeRange.isDateTimeSet()) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    DateTime startDateTime = new DateTime(timeRange.convertStartTimeToString());
+    DateTime startDateTimeWithShift = new DateTime(startDateTime.getValue() + timeRange.getShift());
+    DateTime endDateTime = new DateTime(timeRange.convertEndTimeToString());
+    DateTime endDateTimeWithShift = new DateTime(endDateTime.getValue() + timeRange.getShift());
+    BatchRequest batchRequest = new BatchRequest(flow.getTransport(), flow.getRequestInitializer())
+        .setBatchUrl(new GenericUrl("https://www.googleapis.com/batch/calendar/v3"));
+    String ownerId = eventStorageObject.getEvent(eventId).getOwnerID();
+
+    if (userService.getCurrentUser() == null || !userService.getCurrentUser().getUserId().equals(ownerId)) {
+      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+      return;
+    }
+
+    List<String> joinedParticipantsIds = eventStorageObject.getEvent(eventId).getJoinedIDs();
+    if (!joinedParticipantsIds.contains(ownerId)) {
+      joinedParticipantsIds.add(ownerId);
+    }
+
+    List<Credential> usersCredentials;
+    try {
+      usersCredentials = getCredentialsFromUserList(joinedParticipantsIds);
+    } catch (InterruptedException e2) {
+      e2.printStackTrace();
+      response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return;
+    }
+
+    usersCredentials.forEach(userCredential -> {
+      try {
+        createAndAddBatchRequest(batchRequest, callback, userCredential, startDateTimeWithShift, endDateTimeWithShift);
+        batchRequest.execute();
```

Comment:
Why do you execute it directly? Then you don't batch it at all but only executing a single request each time, if I understand it correctly.

Reply:
You are right, will change